tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc

```diff
@@ -275,1817 +314,2459 @@ TEST(LinalgBackendEigen, add_vector)
-	for (float64_t i = 0; i < 9; ++i)
-	{
-		A[i] = i * 0.5 - 0.5;
-		B[i] = i;
-	}
+    for (index_t i = 0; i < nrows * ncols; ++i)
+        mat[i] = i;
 
-	multiply_by_rectified_linear_derivative(A, B);
+    SGVector<ST> result = colwise_sum(mat);
 
-	for (index_t i = 0; i < 9; ++i)
-		EXPECT_NEAR(i * (A[i] != 0), B[i], 1e-15);
+    for (index_t j = 0; j < ncols; ++j)
+    {
+        ST sum = 0;
+        for (index_t i = 0; i < nrows; ++i)
+            sum += mat(i, j);
+        EXPECT_NEAR(sum, result[j], 1E-15);
+    }
 }
 
-TEST(LinalgBackendEigen, SGVector_norm)
+template <typename ST>
+void check_SGMatrix_colwise_sum_no_diag()
 {
-	const index_t n = 5;
-	SGVector<float64_t> v(n);
-	float64_t gt = 0;
-	for (index_t i = 0; i < n; ++i)
-	{
-		v[i] = i;
-		gt += i * i;
-	}
-	gt = std::sqrt(gt);
+    const index_t nrows = 2, ncols = 3;
+    SGMatrix<ST> mat(nrows, ncols);
 
-	auto result = norm(v);
+    for (index_t i = 0; i < nrows * ncols; ++i)
+        mat[i] = i;
 
-	EXPECT_NEAR(result, gt, 1E-15);
+    SGVector<ST> result = colwise_sum(mat, true);
+
+    EXPECT_NEAR(result[0], 1, 1E-15);
+    EXPECT_NEAR(result[1], 2, 1E-15);
+    EXPECT_NEAR(result[2], 9, 1E-15);
 }
 
-TEST(LinalgBackendEigen, SGVector_qr_solver)
+template <typename ST>
+void check_SGMatrix_block_colwise_sum()
 {
-	const index_t n = 3;
-	float64_t data_A[] = {0.02800922, 0.99326012, 0.15204902,
-	                      0.30492837, 0.39708534, 0.40466969,
-	                      0.36415317, 0.04407589, 0.9095746};
-	float64_t data_b[] = {0.39461571, 0.6816856, 0.43323709};
-	float64_t result[] = {0.07135206, 1.56393127, -0.23141312};
+    const index_t nrows = 2, ncols = 3;
+    SGMatrix<float64_t> mat(nrows, ncols);
 
-	SGMatrix<float64_t> A(data_A, n, n, false);
-	SGVector<float64_t> b(data_b, n, false);
+    for (index_t i = 0; i < nrows; ++i)
+        for (index_t j = 0; j < ncols; ++j)
+            mat(i, j) = i * 10 + j + 1;
 
-	auto x = qr_solver(A, b);
+    auto result = colwise_sum(linalg::block(mat, 0, 0, 2, 3));
 
-	for (index_t i = 0; i < x.size(); ++i)
-		EXPECT_NEAR(x[i], result[i], 1E-7);
+    for (index_t j = 0; j < ncols; ++j)
+    {
+        ST sum = 0;
+        for (index_t i = 0; i < nrows; ++i)
+            sum += mat(i, j);
+        EXPECT_NEAR(sum, result[j], 1E-15);
+    }
 }
 
-TEST(LinalgBackendEigen, SGMatrix_qr_solver)
+
+template <typename ST>
+void check_SGMatrix_rowwise_sum()
 {
-	const index_t n = 3, m = 2;
-	float64_t data_A[] = {0.02800922, 0.99326012, 0.15204902,
-	                      0.30492837, 0.39708534, 0.40466969,
-	                      0.36415317, 0.04407589, 0.9095746};
-	float64_t data_B[] = {0.76775073, 0.88471312, 0.34795225,
-	                      0.94311546, 0.59630347, 0.65820143};
-	float64_t result[] = {-0.73834587, 4.22750496, -1.37484721,
-	                      -1.14718091, 4.49142548, -1.08282992};
+    const index_t nrows = 2, ncols = 3;
+    SGMatrix<ST> mat(nrows, ncols);
 
-	SGMatrix<float64_t> A(data_A, n, n, fa
```

Comment:
not sure the method is even defined for int types?
Generally, we try to avoid TODOs in the codebase (they never get addressed ;)

Reply:
does a call to this method with int type even compile?

Reply:
Yes, this compiles, but I don't think it might not make much sense...

https://github.com/shogun-toolbox/shogun/blob/f6efd89467a7016be8a50bf3f7c0fdf32b075369/src/shogun/mathematics/linalg/LinalgBackendBase.h#L255-L262

Reply:
well as long as the eigen3 call goes through it makes sense to offer it :)
But I am surprised ...